File: go/gofind/pkg/gofind/finder.go
package gofind

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// Finder - the find executor
type Finder struct {
	Settings           *FindSettings
	fileTypes          *FileTypes
	fileResults        *FileResults
	errors             []error
	addResultChan      chan *FileResult
	addResultsDoneChan chan bool
	findDoneChan       chan bool
	errChan            chan error
}

func NewFinder(settings *FindSettings) (*Finder, error) {
	fileTypes, err := FileTypesFromJson()
	if err != nil {
		return nil, err
	}
	finder := &Finder{
		settings,               // Settings
		fileTypes,              // fileTypes
		NewFileResults(),       // fileResults
		[]error{},              // errors
		make(chan *FileResult), // addResultChan
		make(chan bool),        // addResultsDoneChan
		make(chan bool, 1),     // findDoneChan
		make(chan error, 1),    // errChan
	}
	err = finder.validateSettings()
	if err != nil {
		return nil, err
	}
	return finder, nil
}

func (f *Finder) validateSettings() error {
	if len(f.Settings.Paths()) < 1 {
		return fmt.Errorf(StartPathNotDefined)
	}

	for _, p := range f.Settings.Paths() {
		fi, err := os.Lstat(ExpandPath(p))
		if err != nil {
			if os.IsNotExist(err) {
				return fmt.Errorf(StartPathNotFound)
			}
			if os.IsPermission(err) {
				return fmt.Errorf(StartPathNotReadable)
			}
			return err
		}

		if fi.Mode()&os.ModeSymlink == os.ModeSymlink {
			if !f.Settings.FollowSymlinks() {
				return fmt.Errorf(StartPathNotMatchFindSettings)
			}
		} else if fi.IsDir() {
			if !f.isTraversableDirPath(p) {
				return fmt.Errorf(StartPathNotMatchFindSettings)
			}

		} else if fi.Mode().Type().IsRegular() {
			fileResult := f.filterToFileResult(p, fi)
			if fileResult == nil {
				return fmt.Errorf(StartPathNotMatchFindSettings)
			}

		} else {
			return fmt.Errorf(StartPathNotMatchFindSettings)
		}
	}

	if f.Settings.maxDepth > -1 && f.Settings.maxDepth < f.Settings.minDepth {
		return fmt.Errorf(InvalidRangeForMinDepthAndMaxDepth)
	}
	if !f.Settings.maxLastMod.IsZero() && f.Settings.minLastMod.After(f.Settings.maxLastMod) {
		return fmt.Errorf(InvalidRangeForMinLastModAndMaxLastMod)
	}
	if f.Settings.maxSize > 0 && f.Settings.maxSize < f.Settings.minSize {
		return fmt.Errorf(InvalidRangeForMinSizeAndMaxSize)
	}

	return nil
}

func emptyOrMatchesAnyPattern(s string, patterns *Patterns) bool {
	return patterns.IsEmpty() || patterns.MatchesAny(s)
}

func emptyOrNotMatchesAnyPattern(s string, patterns *Patterns) bool {
	return patterns.IsEmpty() || !patterns.MatchesAny(s)
}

func emptyOrAnyMatchesAnyPattern(strings []string, patterns *Patterns) bool {
	return patterns.IsEmpty() || patterns.AnyMatchesAny(strings)
}

func emptyOrNotAnyMatchesAnyPattern(strings []string, patterns *Patterns) bool {
	return patterns.IsEmpty() || !patterns.AnyMatchesAny(strings)
}

func emptyOrMatchesAnyString(s string, strings []string) bool {
	return len(strings) == 0 || Contains(strings, s)
}

func emptyOrNotMatchesAnyString(s string, strings []string) bool {
	return len(strings) == 0 || !Contains(strings, s)
}

func emptyOrMatchesAnyFileType(fileType FileType, fileTypes []FileType) bool {
	return len(fileTypes) == 0 || ContainsFileType(fileTypes, fileType)
}

func emptyOrNotMatchesAnyFileType(fileType FileType, fileTypes []FileType) bool {
	return len(fileTypes) == 0 || !ContainsFileType(fileTypes, fileType)
}

func (f *Finder) isMatchingDirPathByHidden(dirPath string) bool {
	return f.Settings.IncludeHidden() || !IsHiddenPath(dirPath)
}

func (f *Finder) isMatchingDirPathByInPatterns(dirPath string) bool {
	return emptyOrAnyMatchesAnyPattern(GetPathElems(dirPath), f.Settings.InDirPatterns())
}

func (f *Finder) isMatchingDirPathByOutPatterns(dirPath string) bool {
	return emptyOrNotAnyMatchesAnyPattern(GetPathElems(dirPath), f.Settings.OutDirPatterns())
}

func (f *Finder) isTraversableDirPath(dirPath string) bool {
	return f.isMatchingDirPathByHidden(dirPath) &&
		f.isMatchingDirPathByOutPatterns(dirPath)
}

func (f *Finder) isMatchingDirPath(dirPath string) bool {
	return f.isMatchingDirPathByHidden(dirPath) &&
		f.isMatchingDirPathByInPatterns(dirPath) &&
		f.isMatchingDirPathByOutPatterns(dirPath)
}

func (f *Finder) isNullOrMatchingDirPath(dirPath string) bool {
	return strings.TrimSpace(dirPath) == "" ||
		(f.isMatchingDirPathByHidden(dirPath) &&
			f.isMatchingDirPathByInPatterns(dirPath) &&
			f.isMatchingDirPathByOutPatterns(dirPath))
}

func (f *Finder) isMatchingFileNameByHidden(fileName string) bool {
	return f.Settings.IncludeHidden() || !IsHiddenName(fileName)
}

func (f *Finder) isMatchingArchiveExtension(ext string) bool {
	return emptyOrMatchesAnyString(ext, f.Settings.InArchiveExtensions()) &&
		emptyOrNotMatchesAnyString(ext, f.Settings.OutArchiveExtensions())
}

func (f *Finder) hasMatchingArchiveExtension(filePath string) bool {
	if len(f.Settings.InArchiveExtensions()) > 0 || len(f.Settings.OutArchiveExtensions()) > 0 {
		ext := GetExtension(filePath)
		return f.isMatchingArchiveExtension(ext)
	}
	return true
}

func (f *Finder) isMatchingArchiveFileName(fileName string) bool {
	return f.isMatchingFileNameByHidden(fileName) &&
		emptyOrMatchesAnyPattern(fileName, f.Settings.InArchiveFilePatterns()) &&
		emptyOrNotMatchesAnyPattern(fileName, f.Settings.OutArchiveFilePatterns())
}

func (f *Finder) hasMatchingArchiveFileName(filePath string) bool {
	if !f.Settings.InArchiveFilePatterns().IsEmpty() || !f.Settings.OutArchiveFilePatterns().IsEmpty() {
		return f.isMatchingArchiveFileName(filepath.Base(filePath))
	}
	return true
}

func (f *Finder) isMatchingArchiveFilePath(filePath string) bool {
	return f.hasMatchingArchiveExtension(filePath) &&
		f.hasMatchingArchiveFileName(filePath)
}

func (f *Finder) isMatchingArchiveFileResult(fr *FileResult) bool {
	return f.hasMatchingArchiveExtension(fr.FilePath) &&
		f.hasMatchingArchiveFileName(fr.FilePath)
}

func (f *Finder) isMatchingExtension(ext string) bool {
	return emptyOrMatchesAnyString(ext, f.Settings.InExtensions()) &&
		emptyOrNotMatchesAnyString(ext, f.Settings.OutExtensions())
}

func (f *Finder) hasMatchingExtension(filePath string) bool {
	if len(f.Settings.InExtensions()) > 0 || len(f.Settings.OutExtensions()) > 0 {
		ext := GetExtension(filePath)
		return f.isMatchingExtension(ext)
	}
	return true
}

func (f *Finder) isMatchingFileName(fileName string) bool {
	return f.isMatchingFileNameByHidden(fileName) &&
		emptyOrMatchesAnyPattern(fileName, f.Settings.InFilePatterns()) &&
		emptyOrNotMatchesAnyPattern(fileName, f.Settings.OutFilePatterns())
}

func (f *Finder) hasMatchingFileName(filePath string) bool {
	if !f.Settings.InFilePatterns().IsEmpty() || !f.Settings.OutFilePatterns().IsEmpty() {
		return f.isMatchingFileName(filepath.Base(filePath))
	}
	return true
}

func (f *Finder) isMatchingFilePath(filePath string) bool {
	return f.isNullOrMatchingDirPath(filepath.Dir(filePath)) &&
		f.hasMatchingExtension(filePath) &&
		f.hasMatchingFileName(filePath)
}

func (f *Finder) isMatchingFileType(fileType FileType) bool {
	return emptyOrMatchesAnyFileType(fileType, f.Settings.InFileTypes()) &&
		emptyOrNotMatchesAnyFileType(fileType, f.Settings.OutFileTypes())
}

func (f *Finder) isMatchingFileSize(fileSize int64) bool {
	return (f.Settings.maxSize == 0 || fileSize <= f.Settings.maxSize) &&
		(f.Settings.minSize == 0 || fileSize >= f.Settings.minSize)
}

func (f *Finder) isMatchingLastMod(lastMod time.Time) bool {
	return (f.Settings.maxLastMod.IsZero() || lastMod.Compare(f.Settings.maxLastMod) <= 0) &&
		(f.Settings.minLastMod.IsZero() || lastMod.Compare(f.Settings.minLastMod) >= 0)
}

func (f *Finder) isMatchingFileResult(fr *FileResult) bool {
	return f.isMatchingFilePath(fr.FilePath) &&
		f.isMatchingFileType(fr.FileType) &&
		f.isMatchingFileSize(fr.FileSize) &&
		f.isMatchingLastMod(fr.LastMod)
}

func (f *Finder) filterArchiveFilePathToFileResult(filePath string, fi os.FileInfo) *FileResult {
	if !f.Settings.includeArchives && !f.Settings.archivesOnly {
		return nil
	}

	if !f.isMatchingArchiveFilePath(filePath) {
		return nil
	}

	var fileSize int64 = 0
	lastMod := time.Time{}
	if fi != nil {
		fileSize = fi.Size()
		lastMod = fi.ModTime()
	}

	return NewFileResult(filePath, FileTypeArchive, fileSize, lastMod)
}

func (f *Finder) filterRegFilePathToFileResult(filePath string, fileType FileType, fi os.FileInfo) *FileResult {
	if f.Settings.archivesOnly {
		return nil
	}

	if !f.isMatchingFilePath(filePath) || !f.isMatchingFileType(fileType) {
		return nil
	}

	var fileSize int64 = 0
	lastMod := time.Time{}
	if fi != nil {
		fileSize = fi.Size()
		lastMod = fi.ModTime()
	}

	return NewFileResult(filePath, fileType, fileSize, lastMod)
}

func (f *Finder) filterToFileResult(filePath string, fi os.FileInfo) *FileResult {
	if !f.isNullOrMatchingDirPath(filepath.Dir(filePath)) || !f.isMatchingFileNameByHidden(filepath.Base(filePath)) {
		return nil
	}

	fileType := f.fileTypes.GetFileType(filePath)
	if fileType == FileTypeArchive {
		return f.filterArchiveFilePathToFileResult(filePath, fi)
	}

	return f.filterRegFilePathToFileResult(filePath, fileType, fi)
}

func (f *Finder) checkAddFileResult(filePath string, fi os.FileInfo) {
	fileResult := f.filterToFileResult(filePath, fi)
	if fileResult != nil {
		f.addResultChan <- fileResult
	}
}

// this method passed to the filepath.WalkDir method, it must have this signature
func (f *Finder) checkAddFindWalkFile(filePath string, entry fs.DirEntry, err error) error {
	if err != nil {
		fmt.Printf("an error occurred accessing path %q: %v\n", filePath, err)
		return err
	}
	if entry.IsDir() && (!f.isTraversableDirPath(entry.Name())) {
		return filepath.SkipDir
	} else if entry.Type().IsRegular() {
		fi, err := entry.Info()
		if err != nil {
			return err
		}
		f.checkAddFileResult(filePath, fi)
	}
	return nil
}

// if we get true from addResultsDoneChan, set addResultsDone to true
// if we get a result from addResultChan channel, add to fileResults
// if we get an error from errChan channel, add to errors
func (f *Finder) activateFindChannels() {
	addResultsDone := false
	for !addResultsDone {
		select {
		case b := <-f.addResultsDoneChan:
			addResultsDone = b
			f.findDoneChan <- b
		case r := <-f.addResultChan:
			f.fileResults.AddResult(r)
		case e := <-f.errChan:
			f.errors = append(f.errors, e)
		}
	}
}

func (f *Finder) recSetFileResultsForPath(path string, minDepth int, maxDepth int, currentDepth int) error {
	recurse := true
	if currentDepth == maxDepth {
		recurse = false
	} else if maxDepth > -1 && currentDepth > maxDepth {
		return nil
	}
	var dirs []string
	entries, err := os.ReadDir(path)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		fileInfo, _ := entry.Info()
		if fileInfo.Mode()&os.ModeSymlink == os.ModeSymlink {
			if f.Settings.followSymlinks {
				symlinkFilePath := filepath.Join(path, entry.Name())
				resolvedPath, err := filepath.EvalSymlinks(symlinkFilePath)
				if err != nil {
					return err
				}
				fileInfo, err = os.Stat(resolvedPath)
				if err != nil {
					return err
				}
			} else {
				continue
			}
		}
		if fileInfo.IsDir() && recurse && f.isTraversableDirPath(entry.Name()) {
			dirs = append(dirs, filepath.Join(path, entry.Name()))
		} else if fileInfo.Mode().IsRegular() && (minDepth < 0 || currentDepth >= minDepth) {
			f.checkAddFileResult(filepath.Join(path, entry.Name()), fileInfo)
		}
	}
	for _, dir := range dirs {
		err = f.recSetFileResultsForPath(dir, minDepth, maxDepth, currentDepth+1)
		if err != nil {
			return err
		}
	}
	return nil
}

func (f *Finder) setFileResultsForPath(path string) error {
	fi, err := os.Stat(path)
	if err != nil {
		path = ExpandPath(path)
		fi, err = os.Stat(path)
		if err != nil {
			return err
		}
	}
	if fi.IsDir() {
		// if MaxDepth is zero, we can skip since a directory cannot be a result
		if f.Settings.MaxDepth() == 0 {
			return nil
		}

		if f.isTraversableDirPath(path) {
			maxDepth := f.Settings.MaxDepth()
			if !f.Settings.Recursive() {
				maxDepth = 1
			}
			return f.recSetFileResultsForPath(path, f.Settings.MinDepth(), maxDepth, 1)
		}

		return fmt.Errorf(StartPathNotMatchFindSettings)

	} else if fi.Mode().Type().IsRegular() {
		if f.Settings.minDepth > 0 {
			return nil
		}
		f.checkAddFileResult(path, fi)
	} else {
		// TODO: handle symlinks, etc.?
	}
	return nil
}

func (f *Finder) setFileResults() error {
	//if f.Settings.Verbose() {
	//	Log("\nBuilding file result list")
	//}

	// start the find channels with goroutine
	go f.activateFindChannels()

	for _, p := range f.Settings.Paths() {
		if err := f.setFileResultsForPath(p); err != nil {
			return err
		}
	}

	f.addResultsDoneChan <- true
	return nil
}

func (f *Finder) Find() (*FileResults, error) {
	// send to the find channels
	if err := f.setFileResults(); err != nil {
		return nil, err
	}

	findDone := false
	for !findDone {
		select {
		case b := <-f.findDoneChan:
			findDone = b
		}
	}

	// sort the FileResults
	f.fileResults.Sort(f.Settings)

	return f.fileResults, nil
}

File: objc/objcfind/Sources/objcfind/Finder.m
#import "common.h"
#import "FileResultSorter.h"
#import "FileUtil.h"
#import "Finder.h"

@implementation Finder

- (instancetype) initWithSettings:(FindSettings*)settings error:(NSError**)error {
    self = [super init];
    if (self) {
        self.fileTypes = [[FileTypes alloc] init];
        self.settings = settings;
        if (![self validateSettings:settings error:error]) {
            return self;
        }
    }
    return self;
}

- (NSStringEncoding) strToEncoding:(NSString*)s {
    NSStringEncoding encoding =
        CFStringConvertEncodingToNSStringEncoding(CFStringConvertIANACharSetNameToEncoding((CFStringRef) s));
    return encoding;
}

- (BOOL) validateSettings:(FindSettings*)settings error:(NSError**)error {
    if (settings == nil) {
        setError(error, @"Settings not defined");
        return false;
    } else if (settings.paths == nil || [settings.paths count] == 0) {
        setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_DEFINED]);
        return false;
    }
    for (NSString *path in settings.paths) {
        NSString *p = path;
        if (![FileUtil exists:p]) {
            p = [FileUtil expandPath:p];
            if (![FileUtil exists:p]) {
                setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_FOUND]);
                return false;
            }
        }
        if (![FileUtil isReadableFile:p]) {
            setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_READABLE]);
            return false;
        }
        NSString* resolvedPath = p;
        if ([FileUtil isSymlink:p]) {
            if (settings.followSymlinks) {
                resolvedPath = [FileUtil getSymlinkTarget:path];
            } else {
                setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_MATCH_FIND_SETTINGS]);
                return false;
            }
        }
        if ([FileUtil isDirectory:resolvedPath]) {
            // still check p and not resolvedPath because p is the name entered
            if (![self isTraversableDirPath:p]) {
                setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_MATCH_FIND_SETTINGS]);
                return false;
            }
        } else if ([FileUtil isReadableFile:resolvedPath]) {
            // still check p and not resolvedPath because p is the name entered
            FileResult *fr = [self filterToFileResult:p error:error];
            if (*error) {
                return false;
            }
            if (fr == nil) {
                setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_MATCH_FIND_SETTINGS]);
                return false;
            }
        }
    }
    if (settings.maxDepth > -1 && settings.maxDepth < settings.minDepth) {
        setError(error, [NSString stringWithUTF8String:INVALID_RANGE_MINDEPTH_MAXDEPTH]);
        return false;
    } else if (settings.maxLastMod != nil && settings.minLastMod != nil && [settings.maxLastMod isEqualToDate:[settings.maxLastMod earlierDate:settings.minLastMod]]) {
        setError(error, [NSString stringWithUTF8String:INVALID_RANGE_MINLASTMOD_MAXLASTMOD]);
        return false;
    } else if (settings.maxSize > 0 && settings.maxSize < settings.minSize) {
        setError(error, [NSString stringWithUTF8String:INVALID_RANGE_MINSIZE_MAXSIZE]);
        return false;
    }
    return true;
}

- (BOOL) matchesAnyPattern:(NSString*)s patterns:(NSArray<Regex*>*)patterns {
    for (Regex *r in patterns) {
        if ([r test:s]) {
            return true;
        }
    }
    return false;
}

- (BOOL) anyMatchesAnyPattern:(NSArray<NSString*>*)ss patterns:(NSArray<Regex*>*)patterns {
    for (NSString *s in ss) {
        if ([self matchesAnyPattern:s patterns:patterns]) {
            return true;
        }
    }
    return false;
}

- (BOOL) emptyOrMatchesAnyPattern:(NSString*)s patterns:(NSArray<Regex*>*)patterns {
    return [patterns count] == 0 || [self matchesAnyPattern:s patterns:patterns];
}

- (BOOL) emptyOrNotMatchesAnyPattern:(NSString*)s patterns:(NSArray<Regex*>*)patterns {
    return [patterns count] == 0 || ![self matchesAnyPattern:s patterns:patterns];
}

- (BOOL) emptyOrAnyMatchesAnyPattern:(NSArray<NSString*>*)ss patterns:(NSArray<Regex*>*)patterns {
    return [patterns count] == 0 || [self anyMatchesAnyPattern:ss patterns:patterns];
}

- (BOOL) emptyOrNotAnyMatchesAnyPattern:(NSArray<NSString*>*)ss patterns:(NSArray<Regex*>*)patterns {
    return [patterns count] == 0 || ![self anyMatchesAnyPattern:ss patterns:patterns];
}

- (BOOL) emptyOrMatchesAnyString:(NSString*)s strings:(NSArray<NSString*>*)strings {
    return [strings count] == 0 || [strings containsObject:s];
}

- (BOOL) emptyOrNotMatchesAnyString:(NSString*)s strings:(NSArray<NSString*>*)strings {
    return [strings count] == 0 || ![strings containsObject:s];
}

- (BOOL) emptyOrMatchesAnyFileType:(FileType)fileType fileTypes:(NSArray<NSNumber*>*)fileTypes {
    if ([fileTypes count] > 0) {
        NSNumber *num = [NSNumber numberWithInt:fileType];
        return [fileTypes containsObject:num];
    }
    return true;
}

- (BOOL) emptyOrNotMatchesAnyFileType:(FileType)fileType fileTypes:(NSArray<NSNumber*>*)fileTypes {
    if ([fileTypes count] > 0) {
        NSNumber *num = [NSNumber numberWithInt:fileType];
        return ![fileTypes containsObject:num];
    }
    return true;
}

- (BOOL) isMatchingDirPathByHidden:(NSString*)dirPath {
    return self.settings.includeHidden || ![FileUtil isHiddenPath:dirPath];
}

- (BOOL) isMatchingDirPathByInPatterns:(NSString*)dirPath {
    return [self emptyOrMatchesAnyPattern:dirPath patterns:self.settings.inDirPatterns];
}

- (BOOL) isMatchingDirPathByOutPatterns:(NSString*)dirPath {
    return [self emptyOrNotMatchesAnyPattern:dirPath patterns:self.settings.outDirPatterns];
}

- (BOOL) isTraversableDirPath:(NSString*)dirPath {
    return [self isMatchingDirPathByHidden:dirPath]
        && [self isMatchingDirPathByOutPatterns:dirPath];
}

- (BOOL) isMatchingDirPath:(NSString*)dirPath {
    return [self isMatchingDirPathByHidden:dirPath]
        && [self isMatchingDirPathByInPatterns:dirPath]
        && [self isMatchingDirPathByOutPatterns:dirPath];
}

- (BOOL) isNullOrMatchingDirPath:(NSString*)dirPath {
    return dirPath == nil
        || ([self isMatchingDirPathByHidden:dirPath]
        && [self isMatchingDirPathByInPatterns:dirPath]
        && [self isMatchingDirPathByOutPatterns:dirPath]);
}

- (BOOL) isMatchingFileNameByHidden:(NSString*)fileName {
    return self.settings.includeHidden || ![FileUtil isHiddenName:fileName];
}

- (BOOL) isMatchingArchiveExtension:(NSString*)ext {
    return [self emptyOrMatchesAnyString:ext strings:self.settings.inArchiveExtensions]
        && [self emptyOrNotMatchesAnyString:ext strings:self.settings.outArchiveExtensions];
}

- (BOOL) isMatchingArchiveExtensionForFilePath:(NSString*)filePath {
    if ([self.settings.inArchiveExtensions count] > 0 || [self.settings.outArchiveExtensions count] > 0) {
        NSString *fileName = [filePath lastPathComponent];
        NSString *ext = [FileUtil getExtension:fileName];
        return [self isMatchingArchiveExtension:ext];
    }
    return true;
}

- (BOOL) isMatchingArchiveFileName:(NSString*)fileName {
    return [self emptyOrMatchesAnyPattern:fileName patterns:self.settings.inArchiveFilePatterns]
        && [self emptyOrNotMatchesAnyPattern:fileName patterns:self.settings.outArchiveFilePatterns];
}

- (BOOL) isMatchingArchiveFileNameForFilePath:(NSString*)filePath {
    if ([self.settings.inArchiveFilePatterns count] > 0 || [self.settings.outArchiveFilePatterns count] > 0) {
        NSString *fileName = [filePath lastPathComponent];
        return [self isMatchingArchiveFileName:fileName];
    }
    return true;
}

- (BOOL) isMatchingArchiveFilePath:(NSString*)filePath {
    return [self isMatchingArchiveExtensionForFilePath:filePath]
        && [self isMatchingArchiveFileNameForFilePath:filePath];
}

- (BOOL) isMatchingArchiveFileResult:(FileResult*)fileResult {
    return [self isMatchingArchiveFilePath:[fileResult filePath]];
}


- (BOOL) isMatchingExtension:(NSString*)ext {
    return [self emptyOrMatchesAnyString:ext strings:self.settings.inExtensions]
        && [self emptyOrNotMatchesAnyString:ext strings:self.settings.outExtensions];
}

- (BOOL) isMatchingExtensionForFilePath:(NSString*)filePath {
    if ([self.settings.inExtensions count] > 0 || [self.settings.outExtensions count] > 0) {
        NSString *fileName = [filePath lastPathComponent];
        NSString *ext = [FileUtil getExtension:fileName];
        return [self isMatchingExtension:ext];
    }
    return true;
}

- (BOOL) isMatchingFileName:(NSString*)fileName {
    return [self emptyOrMatchesAnyPattern:fileName patterns:self.settings.inFilePatterns]
        && [self emptyOrNotMatchesAnyPattern:fileName patterns:self.settings.outFilePatterns];
}

- (BOOL) isMatchingFileNameForFilePath:(NSString*)filePath {
    if ([self.settings.inFilePatterns count] > 0 || [self.settings.outFilePatterns count] > 0) {
        NSString *fileName = [filePath lastPathComponent];
        return [self isMatchingFileName:fileName];
    }
    return true;
}

- (BOOL) isMatchingFilePath:(NSString*)filePath {
    return [self isMatchingExtensionForFilePath:filePath]
        && [self isMatchingFileNameForFilePath:filePath];
}

- (BOOL) isMatchingFileType:(FileType)fileType {
    return [self emptyOrMatchesAnyFileType:fileType fileTypes:[self.settings inFileTypes]]
        && [self emptyOrNotMatchesAnyFileType:fileType fileTypes:[self.settings outFileTypes]];
}

- (BOOL) isMatchingFileSize:(unsigned long long)fileSize {
    if ([self.settings maxSize] > 0 || [self.settings minSize] > 0) {
        NSNumber *numFileSize = [[NSNumber alloc] initWithUnsignedLongLong:fileSize];
        return ([self.settings maxSize] == 0 || [numFileSize longValue] <= (long)[self.settings maxSize])
            && ([self.settings minSize] == 0 || [numFileSize longValue] >= (long)[self.settings minSize]);
    }
    return true;
}

- (BOOL) isMatchingLastMod:(NSDate*)lastMod {
    if ([self.settings maxLastMod] != nil || [self.settings minLastMod] != nil) {
        return lastMod != nil
            && ([self.settings maxLastMod] == nil || [lastMod isLessThanOrEqualTo:[self.settings maxLastMod]])
            && ([self.settings minLastMod] == nil || [lastMod isGreaterThanOrEqualTo:[self.settings minLastMod]]);
    }
    return true;
}

- (BOOL) isMatchingFileResult:(FileResult*)fileResult {
    return [self isMatchingFilePath:[fileResult filePath]]
        && [self isMatchingFileType:[fileResult fileType]]
        && [self isMatchingFileSize:[fileResult fileSize]]
        && [self isMatchingLastMod:[fileResult lastMod]];
}

- (FileResult*) filterArchiveFilePathToFileResult:(NSString*)filePath error:(NSError**)error {
    if (!self.settings.includeArchives && !self.settings.archivesOnly) {
        return nil;
    }
    
    if (![self isMatchingArchiveFilePath:filePath]) {
        return nil;
    }

    unsigned long long fileSize = 0;
    NSDate *lastMod = nil;
    return [[FileResult alloc] initWithFilePath:filePath fileType:FileTypeArchive fileSize:fileSize lastMod:lastMod];
}

- (FileResult*) filterRegularFilePathToFileResult:(NSString*)filePath fileType:(FileType)fileType error:(NSError**)error {
    if (self.settings.archivesOnly) {
        return nil;
    }

    if (![self isMatchingFilePath:filePath] || ![self isMatchingFileType:fileType]) {
        return nil;
    }

    unsigned long long fileSize = 0;
    NSDate *lastMod = nil;
    if ([self.settings needSize] || [self.settings needLastMod]) {
        NSDictionary<NSFileAttributeKey, id> *stat = [FileUtil getFileAttributes:filePath error:error];
        if (*error) {
            return nil;
        }
        if ([self.settings needSize]) fileSize = stat.fileSize;
        if ([self.settings needLastMod]) lastMod = stat.fileModificationDate;

        if (![self isMatchingFileSize:fileSize] || ![self isMatchingLastMod:lastMod]) {
            return nil;
        }
    }

    return [[FileResult alloc] initWithFilePath:filePath fileType:fileType fileSize:fileSize lastMod:lastMod];
}

- (FileResult*) filterToFileResult:(NSString*)filePath error:(NSError**)error {
    NSString *parent = [filePath stringByDeletingLastPathComponent];
    if (![self isNullOrMatchingDirPath:parent]) {
        return nil;
    }

    NSString *fileName = [filePath lastPathComponent];
    if (![self isMatchingFileNameByHidden:fileName]) {
        return nil;
    }

    FileType fileType = [self.fileTypes getFileType:fileName];
    if (fileType == FileTypeArchive) {
        return [self filterArchiveFilePathToFileResult:filePath error:error];
    }

    return [self filterRegularFilePathToFileResult:filePath fileType:fileType error:error];
}

- (NSArray<FileResult*>*) recGetFileResults:(NSString*)dirPath minDepth:(long)minDepth maxDepth:(long)maxDepth currentDepth:(long)currentDepth error:(NSError**)error {
    NSMutableArray *fileResults = [NSMutableArray array];
    BOOL recurse = true;
    if (currentDepth == maxDepth) {
        recurse = false;
    } else if (maxDepth > -1 && currentDepth > maxDepth) {
        return fileResults;
    }
    
    NSArray<NSString*>* dirElems = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:dirPath
                                                                             error:error];
    if (*error) {
        return nil;
    }

    NSMutableArray *subDirs = [NSMutableArray array];
    for (NSString *dirElem in dirElems) {
        NSString* subPath = [FileUtil joinPath:dirPath childPath:dirElem];
        BOOL linkIsDir = false;
        BOOL linkIsFile = false;
        if ([FileUtil isSymlink:subPath]) {
            if (self.settings.followSymlinks) {
                NSString* targetPath = [FileUtil getSymlinkTarget:subPath];
                if (targetPath != nil) {
                    if ([FileUtil isDirectory:targetPath]) {
                        linkIsDir = true;
                    } else if ([FileUtil isReadableFile:targetPath]) {
                        linkIsFile = true;
                    }
                }
            } else {
                continue;
            }
        }
        if ([FileUtil isDirectory:subPath] || linkIsDir) {
            if (recurse && [self isTraversableDirPath:dirElem]) {
                [subDirs addObject:subPath];
            }
        } else if (([FileUtil isReadableFile:subPath] || linkIsFile) && (minDepth < 0 || currentDepth >= minDepth)) {
            FileResult *fileResult = [self filterToFileResult:subPath error:error];
            if (*error) {
                return nil;
            }
            if (fileResult != nil) {
                [fileResults addObject:fileResult];
            }
        }
    }
    
    for (NSString *subDir in subDirs) {
        NSArray<FileResult*> *dirResults = [self recGetFileResults:subDir minDepth:minDepth maxDepth:maxDepth currentDepth:(currentDepth + 1) error:error];
        if (*error) {
            return nil;
        }
        [fileResults addObjectsFromArray:dirResults];
    }

    return [NSArray arrayWithArray:fileResults];
}

- (NSArray<FileResult*>*) getFileResults:(NSString*)filePath error:(NSError**)error {
    NSString* fp = filePath;
    if (![FileUtil exists:filePath]) {
        fp = [FileUtil expandPath:filePath];
        if (![FileUtil exists:fp]) {
            setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_FOUND]);
            return nil;
        }
    }
    NSMutableArray *fileResults = [NSMutableArray array];
    BOOL linkIsDir = false;
    BOOL linkIsFile = false;
    if ([FileUtil isSymlink:fp]) {
        if (self.settings.followSymlinks) {
            NSString* targetPath = [FileUtil getSymlinkTarget:fp];
            if (targetPath != nil) {
                if ([FileUtil isDirectory:targetPath]) {
                    linkIsDir = true;
                } else if ([FileUtil isReadableFile:targetPath]) {
                    linkIsFile = true;
                }
            }
        } else {
            setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_MATCH_FIND_SETTINGS]);
            return nil;
        }
    }
    if ([FileUtil isDirectory:fp] || linkIsDir) {
        // if maxDepth is zero, we can skip since a directory cannot be a result
        if (self.settings.maxDepth == 0) {
            return fileResults;
        }
        if ([self isTraversableDirPath:fp]) {
            long maxDepth = self.settings.recursive ? self.settings.maxDepth : 1;
            return [self recGetFileResults:fp minDepth:self.settings.minDepth maxDepth:maxDepth currentDepth:1 error:error];
        } else {
            setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_MATCH_FIND_SETTINGS]);
            return nil;
        }
    } else if ([FileUtil isReadableFile:fp] || linkIsFile) {
        // if minDepth > zero, we can skip since the file is at depth zero
        if (self.settings.minDepth > 0) {
            return fileResults;
        }
        FileResult *fileResult = [self filterToFileResult:fp error:error];
        if (*error) {
            return nil;
        }
        if (fileResult != nil) {
            [fileResults addObject:fileResult];
        } else {
            setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_MATCH_FIND_SETTINGS]);
            return nil;
        }
    } else {
        setError(error, [NSString stringWithUTF8String:STARTPATH_NOT_MATCH_FIND_SETTINGS]);
        return nil;
    }

    return [NSArray arrayWithArray:fileResults];
}

- (NSArray<FileResult*>*) find:(NSError**)error {
    NSMutableArray<FileResult*> *fileResults = [NSMutableArray array];
    for (NSString *p in self.settings.paths) {
        NSArray<FileResult*> *pathResults = [self getFileResults:p error:error];
        if (*error) {
            return nil;
        }
        [fileResults addObjectsFromArray:pathResults];
    }

    if ([fileResults count] > 1) {
        FileResultSorter *fileResultSorter = [[FileResultSorter alloc] initWithSettings:self.settings];
        return [fileResultSorter sort:[NSArray arrayWithArray:fileResults]];
    }
    return fileResults;
}

- (NSArray<NSString*>*) getMatchingDirs:(NSArray<FileResult*>*)fileResults {
    NSMutableSet<NSString*> *dirSet = [NSMutableSet set];
    for (FileResult *fr in fileResults) {
        [dirSet addObject:[[fr description] stringByDeletingLastPathComponent]];
    }
    NSArray *dirArr = [NSArray arrayWithArray:[dirSet allObjects]];
    return [dirArr sortedArrayUsingComparator:^NSComparisonResult(NSString *s1, NSString *s2) {
        return [s1 compare:s2];
    }];
}

- (void) printMatchingDirs:(NSArray<FileResult*>*)fileResults formatter:(FileResultFormatter*)formatter {
    NSArray<NSString*> *dirs = [self getMatchingDirs:fileResults];
    if ([dirs count] > 0) {
        logMsg([NSString stringWithFormat:@"\nMatching directories (%lu):", [dirs count]]);
        for (NSString *d in dirs) {
            logMsg(formatter.formatDirPath(d));
        }
    } else {
        logMsg(@"\nMatching directories: 0");
    }
}

- (void) printMatchingFiles:(NSArray<FileResult*>*)fileResults formatter:(FileResultFormatter*)formatter {
    if ([fileResults count] > 0) {
        logMsg([NSString stringWithFormat:@"\nMatching files (%lu):", [fileResults count]]);
        for (FileResult *fr in fileResults) {
            logMsg([formatter formatFileResult:fr]);
        }
    } else {
        logMsg(@"\nMatching files: 0");
    }
}

@end
